// frontity/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDerived = exports.Slot = exports.useFills = exports.URL = exports.decode = exports.fetch = exports.createStore = exports.connect = exports.warn = exports.error = exports.useConnect = exports.batch = exports.unobserve = exports.observe = exports.loadable = exports.HelmetProvider = exports.Head = exports.styled = exports.CacheProvider = exports.keyframes = exports.Global = exports.css = void 0;
require("@frontity/type-declarations");
var react_1 = require("@emotion/react");
Object.defineProperty(exports, "css", { enumerable: true, get: function () { return react_1.css; } });
Object.defineProperty(exports, "Global", { enumerable: true, get: function () { return react_1.Global; } });
Object.defineProperty(exports, "keyframes", { enumerable: true, get: function () { return react_1.keyframes; } });
Object.defineProperty(exports, "CacheProvider", { enumerable: true, get: function () { return react_1.CacheProvider; } });
var styled_1 = require("@emotion/styled");
Object.defineProperty(exports, "styled", { enumerable: true, get: function () { return __importDefault(styled_1).default; } });
var react_helmet_async_1 = require("react-helmet-async");
Object.defineProperty(exports, "Head", { enumerable: true, get: function () { return react_helmet_async_1.Helmet; } });
Object.defineProperty(exports, "HelmetProvider", { enumerable: true, get: function () { return react_helmet_async_1.HelmetProvider; } });
var component_1 = require("@loadable/component");
Object.defineProperty(exports, "loadable", { enumerable: true, get: function () { return __importDefault(component_1).default; } });
var connect_1 = require("@frontity/connect");
Object.defineProperty(exports, "observe", { enumerable: true, get: function () { return connect_1.observe; } });
Object.defineProperty(exports, "unobserve", { enumerable: true, get: function () { return connect_1.unobserve; } });
Object.defineProperty(exports, "batch", { enumerable: true, get: function () { return connect_1.batch; } });
Object.defineProperty(exports, "useConnect", { enumerable: true, get: function () { return connect_1.useConnect; } });
var error_1 = require("@frontity/error");
Object.defineProperty(exports, "error", { enumerable: true, get: function () { return error_1.error; } });
Object.defineProperty(exports, "warn", { enumerable: true, get: function () { return error_1.warn; } });
const connect_2 = __importStar(require("@frontity/connect"));
exports.connect = connect_2.default;
exports.createStore = connect_2.createStore;
const node_fetch_1 = __importDefault(require("node-fetch"));
exports.fetch = node_fetch_1.default;
const client_1 = __importDefault(require("./utils/decode/client"));
const server_1 = __importDefault(require("./utils/decode/server"));
exports.decode = typeof window !== "undefined" ? client_1.default : server_1.default;
var url_1 = require("./utils/url");
Object.defineProperty(exports, "URL", { enumerable: true, get: function () { return __importDefault(url_1).default; } });
var use_fills_1 = require("./utils/slot-and-fill/use-fills");
Object.defineProperty(exports, "useFills", { enumerable: true, get: function () { return __importDefault(use_fills_1).default; } });
var slot_1 = require("./utils/slot-and-fill/slot");
Object.defineProperty(exports, "Slot", { enumerable: true, get: function () { return __importDefault(slot_1).default; } });
var is_derived_1 = require("./utils/is-derived");
Object.defineProperty(exports, "isDerived", { enumerable: true, get: function () { return __importDefault(is_derived_1).default; } });
